src/common/services/session/session.service.js

```diff
@@ -110,28 +116,50 @@ function session( $cookies, $rootScope, $http, envService ) {
 
   /* Private Methods */
   function updateCurrentSession( encoded_value ) {
-    var cortexSession = {};
+    let cortexSession = {};
     if ( angular.isDefined( encoded_value ) ) {
       cortexSession = jwtDecode( encoded_value );
     }
+    // Set expiration timeout
+    setSessionTimeout();
     // Copy new session into current session object
     angular.copy( cortexSession, session );
     // Update sessionSubject with new value
     sessionSubject.next( session );
   }
 
+  function setSessionTimeout() {
+    let encodedSession = $cookies.get( 'give-session' );
+    // Cancel current session timeout
+    if ( angular.isDefined( sessionTimeout ) ) {
+      $timeout.cancel( sessionTimeout );
+      sessionTimeout = undefined;
+    }
+    // Decode give-session cookie and set timeout based on expiration
+    if ( angular.isDefined( encodedSession ) ) {
+      let giveSession = jwtDecode( encodedSession ),
+        timeout = new Date( giveSession.exp * 1000 ) - Date.now();
+      if ( timeout > 0 ) {
+        sessionTimeout = $timeout( timeout );
+        sessionTimeout.then( () => {
+          updateCurrentSession( $cookies.get( 'cortex-session' ) );
```

Comment:
When this timeout has occurred, do we need to manually change the user's role from REGISTERED to IDENTIFIED? This doesn't fire off a cortex request that will come back with a different cookie so the cookie will still contain REGISTERED. This will be read by `sessionService.getRole()` in your enforcer service and the user will still look registered. Does this sound right or am I missing something?